## Homework 3
This homework asks you to scrape from three different sources, please follow the instructions and do the best you can. With the exception of the first cell that imports the URL request and the beautiful soup library, I have not pre-written any code for you. Do not be afraid--you may look at the tutorial for examples, as well as the Beautiful Soup documentation, and any other Python resource (such a Stack overflow) if need be. 

This homework is intended to be challenging, if you only get 70% of the stuff done than you did very well.

In [133]:
from urllib.request import urlopen
from bs4 import BeautifulSoup

## Shakespeare 
I've posted some relatively simple HTML of the first act of William Shakespeare's The Tempest here: http://floatingmedia.com/columbia/tempest.html please use that URL to download the HTML and put it through Beautiful Soup. (It also may be very helpful for you to go to that page in chrome and inspect elements why you do this.) 
**Please note: there are 2 scenes in Act 1**



In the cells below you should you make two lines--a variable that requests and reads the HTML from the URL (http://floatingmedia.com/columbia/tempest.html), and a second variable that passes the raw HTML into Beautiful Soup. You will use that beautiful soup variable to search the HTML.

In [134]:
my_url = "http://floatingmedia.com/columbia/tempest.html"
raw_html = urlopen(my_url).read()
soup_doc = BeautifulSoup(raw_html, "html.parser")

Get the title of the play:

In [135]:
soup_doc.title

<title>The Tempest</title>

Get the HTML that contains the setting of Act One Scene 1:

In [136]:
Act_1_Scene1 = soup_doc.find_all("div")[1]
Act_1_Scene1

<div>
<p class="stagedir"> a tempestuous noise of thunder and lightning heard.</p>
<p class="stagedir">[Enter a Master and a Boatswain]</p>
<ul><li class="playtext"><strong>Master. </strong>Boatswain!
</li></ul>
<ul><li class="playtext"><strong>Boatswain. </strong>Here, master: what cheer?
</li></ul>
<ul><li class="playtext"><strong>Master. </strong>Good, speak to the mariners: fall to't, yarely,
 <span class="playlinenum">5</span><br/>or we run ourselves aground: bestir, bestir.
</li></ul>
<p class="stagedir">[Exit]</p>
<p class="stagedir">[Enter Mariners]</p>
<ul><li class="playtext"><strong>Boatswain. </strong>Heigh, my hearts! cheerly, cheerly, my hearts!
<br/>yare, yare! Take in the topsail. Tend to the
 <span class="playlinenum">10</span><br/>master's whistle. Blow, till thou burst thy wind,
<br/>if room enough!
<br/><i>[Enter ALONSO, SEBASTIAN, ANTONIO, FERDINAND,]</i>
<br/>GONZALO, and others]
</li></ul>
<ul><li class="playtext"><strong>Alonso. </strong>Good boatswain, have car

Get the setting of Act One Scene 1 (as a string):

In [137]:
Act_1_Scene1.find('p').string

' a tempestuous noise of thunder and lightning heard.'

Get the setting of scene 2

In [138]:
Act_1_Scene2_Setting = soup_doc.find_all("div")[2]
Act_1_Scene2_Setting.find_all('p')[1].string

'The island. Before PROSPERO’S cell.'

Get the name of the first character in Scene 1

In [139]:
Act_1_Scene1.find('ul')

<ul><li class="playtext"><strong>Master. </strong>Boatswain!
</li></ul>

In [140]:
Act_1_Scene1.find('strong').string

'Master. '

In [141]:
#Act_1_Scene1.find_all('strong') this is the list you need for for loop.

Get a list of all of characters in Scene 1 (repeats are fine, you need a loop here)

In [142]:
for character in Act_1_Scene1.find_all('strong'):
    Act_1_Scene1.find ('strong')
    print(character.string)

Master. 
Boatswain. 
Master. 
Boatswain. 
Alonso. 
Boatswain. 
Antonio. 
Boatswain. 
Gonzalo. 
Boatswain. 
Gonzalo. 
Boatswain. 
Gonzalo. 
Boatswain. 
Sebastian. 
Boatswain. 
Antonio. 
Gonzalo. 
Boatswain. 
Mariners. 
Boatswain. 
Gonzalo. 
Sebastian. 
Antonio. 
Gonzalo. 
Antonio. 
Sebastian. 
Gonzalo. 


**Bonus:** write a script that takes all the duplicates out of the list. (This is a python thing, not a beautiful soup thing.)

Display every stage direction in _scene 2_

In [143]:
Act_1_Scene2 = soup_doc.find_all("div")[3]
#Act_1_Scene2.find(class_="stagedir").string

In [144]:
for stage in Act_1_Scene2.find_all(class_="stagedir"):
    Act_1_Scene2.find(class_="stagedir")
    print(stage.string)

[Enter PROSPERO and MIRANDA]
[Enter ARIEL]
[Exit]
[Enter CALIBAN]
[Draws, and is charmed from moving]
[Exeunt]


Get the HTML containing Miranda's first speech in Scene 2

In [145]:
Mira_Speech = Act_1_Scene2.find_all('ul')[0]
Mira_Speech

<ul><li class="playtext"><strong>Miranda. </strong>If by your art, my dearest father, you have
 <span class="playlinenum">85</span><br/>Put the wild waters in this roar, allay them.
<br/>The sky, it seems, would pour down stinking pitch,
<br/>But that the sea, mounting to the welkin's cheek,
<br/>Dashes the fire out. O, I have suffered
<br/>With those that I saw suffer: a brave vessel,
 <span class="playlinenum">90</span><br/>Who had, no doubt, some noble creature in her,
<br/>Dash'd all to pieces. O, the cry did knock
<br/>Against my very heart. Poor souls, they perish'd.
<br/>Had I been any god of power, I would
<br/>Have sunk the sea within the earth or ere
 <span class="playlinenum">95</span><br/>It should the good ship so have swallow'd and
<br/>The fraughting souls within her.
</li></ul>

In [146]:
Act_1_Scene2.find_all('li')[0].strong.next.next

'If by your art, my dearest father, you have\n '

Now display those same lines without the speaker's name (need to use siblings and a loop to get this)

In [147]:
for speech in Act_1_Scene2.find('li').find_all('br'):
    print(speech.nextSibling)

Put the wild waters in this roar, allay them.

The sky, it seems, would pour down stinking pitch,

But that the sea, mounting to the welkin's cheek,

Dashes the fire out. O, I have suffered

With those that I saw suffer: a brave vessel,
 
Who had, no doubt, some noble creature in her,

Dash'd all to pieces. O, the cry did knock

Against my very heart. Poor souls, they perish'd.

Had I been any god of power, I would

Have sunk the sea within the earth or ere
 
It should the good ship so have swallow'd and

The fraughting souls within her.



Get the HTML containing the speech after Miranda's

In [148]:
Act_1_Scene2.find_all('ul')[1]
#or
#Act_1_Scene2.find('ul').next_sibling.next_sibling

<ul><li class="playtext"><strong>Prospero. </strong>Be collected:
<br/>No more amazement: tell your piteous heart
<br/>There's no harm done.
 <span class="playlinenum">100</span></li></ul>

## Box Office Mojo 
I've posted some slightly simplified HTML of the current worldwide box office leaders here: http://floatingmedia.com/columbia/boxoffice.htm please use that URL to download the HTML and put it through Beautiful Soup. (It also may be very helpful for you to go to that page in chrome and inspect elements why you do this.) 

For this part of the assignment you should focus mostly on tr (table rows) amd td (table cells) -- and search them using find_all('tr')[1] in order to get a specific row, and find_all('td')[2] to get a specific cell.

Just like with the Shakespeare write two lines get the raw HTML and then transform the raw HTML into a beautiful soup variable

In [149]:
my_url2 = " http://floatingmedia.com/columbia/boxoffice.htm"
raw_html = urlopen(my_url2).read()
soup_doc = BeautifulSoup(raw_html, "html.parser")

Get the text "2017 worldwide grosses"

In [150]:
soup_doc.find('h1')

<h1>2017 WORLDWIDE GROSSES</h1>

Get an HTML list of all of the column names from the header of the table: rank, title, etc...

In [151]:
soup_doc.find('tr')

<tr bgcolor="#dcdcdc"><td align="center">Rank</td><td align="center">Title</td><td align="center">Studio</td><td align="center">Worldwide</td><td align="center">Domestic</td><td align="center">Overseas</td></tr>

Make a variable called "headers" and set it equal to the list of strings of the column names and header. In the next line write: "headers[1]" your output should be "Title"

In [152]:
headers = soup_doc.find('tr').find_all()
headers[1].string

'Title'

Get the HTML for the row that contains the #1 movie

In [153]:
soup_doc.find_all('tr')[1].find_all('td')[1].find('b')

<b>Beauty and the Beast (2017)</b>

Get the HTML for the row that contains the #10 movie 

In [154]:
soup_doc.find_all('tr')[10].find_all('td')[1].find('b')

<b>xXx: The Return of Xander Cage</b>

Get a list of the top 10 films (You'll need to use the loop and specify a range for the loop.) The list should look like this:

`Beauty and the Beast (2017)
The Fate of the Furious
Guardians of the Galaxy Vol. 2
Logan
Kong: Skull Island
Pirates of the Caribbean: Dead Men Tell No Tales
The Boss Baby
Fifty Shades Darker
Your Name.
xXx: The Return of Xander Cage`

In [155]:
for movie in soup_doc.find_all('b')[0:20]:
    print(movie.string)

Beauty and the Beast (2017)
$1,241.3
The Fate of the Furious
$1,229.4
Guardians of the Galaxy Vol. 2
$818.6
Logan
$612.4
Kong: Skull Island
$565.9
Pirates of the Caribbean: Dead Men Tell No Tales
$510.5
The Boss Baby
$483.6
Fifty Shades Darker
$378.8
Your Name.
$354.4
xXx: The Return of Xander Cage
$346.1


Display as strings the top 10 films by name and worldwide gross. Like this:

`Beauty and the Beast (2017) $502.2
The Fate of the Furious $223.9
Guardians of the Galaxy Vol. 2 $356.7
Logan $226.2
Kong: Skull Island $167.8
Pirates of the Caribbean: Dead Men Tell No Tales $117.6
The Boss Baby $171.0
Fifty Shades Darker $114.4
Your Name. $4.9
xXx: The Return of Xander Cage $44.9`

In [156]:
for movie in soup_doc.find_all('b')[0:20]:
    print(movie.string)

Beauty and the Beast (2017)
$1,241.3
The Fate of the Furious
$1,229.4
Guardians of the Galaxy Vol. 2
$818.6
Logan
$612.4
Kong: Skull Island
$565.9
Pirates of the Caribbean: Dead Men Tell No Tales
$510.5
The Boss Baby
$483.6
Fifty Shades Darker
$378.8
Your Name.
$354.4
xXx: The Return of Xander Cage
$346.1


Display the link for the movie `Logan`:

In [157]:
link = soup_doc.find_all('tr')[4].find_all('td')[1].find('a')
get_url = link['href']
print(get_url)

http://www.boxofficemojo.com/movies/?id=wolverine2017.htm


Create a list for `Logan` that looks like this:

`['4', 'Logan', 'Fox', '$612.4', '$226.2', '$386.2']`

In [166]:
Logan=[soup_doc.find_all('tr')[4].find_all('td')]
for element in Logan_list:
    Logan_list = soup_doc.find_all('tr')[4].find_all('td')
    Logan = [element.string]
    print(Logan)
    

['4']
['4']
['Logan']
['Logan']
['Logan']
['Logan']
['Fox']
['Fox']
['$612.4']
['$612.4']
['$612.4']
['$226.2']
['$226.2']
['$386.2']
['$386.2']


Here is some premade code. If you got your headers list and your logan_list right, this will make you a nice dictionary for Logan.

In [159]:
logan_dict = dict(zip(headers, Logan))
print(logan_dict)

{<td align="center">Rank</td>: '$386.2'}


## Advanced Challenge
Pitchfork is no go, so try to get out all of the artists and albums (and synopses) from NPRs 2015 list: http://www.npr.org/2015/12/07/458447213/npr-musics-50-favorite-albums-of-2015
Use as many cells as you would like, and don't forget to make a new URL request and a new beautiful soup variable